pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -100,6 +112,73 @@ func (r *MachineSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 // Reconcile reconciles CAPI and MAPI machines for their respective namespaces.
 func (r *MachineSyncReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", req.Namespace, "name", req.Name)
+
+	logger.V(1).Info("Reconciling machine")
+	defer logger.V(1).Info("Finished reconciling machine")
+
+	var (
+		mapiMachineNotFound bool
+		capiMachineNotFound bool
+	)
+
+	// Get the MAPI Machine
+	mapiMachine := &machinev1beta1.Machine{}
+	if err := r.Get(ctx, req.NamespacedName, mapiMachine); apierrors.IsNotFound(err) {
+		logger.Info("MAPI Machine not found")
+
+		mapiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get MAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("%w: %w", errFailedToGetMAPIMachine, err)
+	}
+
+	// Get the corresponding CAPI Machine
+	capiMachine := &capiv1beta1.Machine{}
+	capiNamespacedName := client.ObjectKey{
+		Namespace: r.CAPINamespace,
+		Name:      req.Name,
+	}
+
+	if err := r.Get(ctx, capiNamespacedName, capiMachine); apierrors.IsNotFound(err) {
+		logger.Info("CAPI Machine not found")
+
+		capiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get CAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("%w: %w", errFailedToGetCAPIMachine, err)
+	}
+
+	if mapiMachineNotFound && capiMachineNotFound {
+		logger.Info("CAPI and MAPI machines not found, nothing to do")
```

Comment:
When do you envision hitting this case? When we get a req for a delete a MAPI/CAPI machine after we just deleted the other counterpart?

Reply:
I'm not entirely sure... I was thinking about deletion, although I mostly threw this in for completeness.

Reply:
Ok 👍